### PR TITLE
Adding SubtractOperation and AverageOperation

### PR DESCRIPTION
**Pull request by uditarora**
- SubtractOperation allows subtraction between two tensors and enables gradient to flow through the subtraction
- AverageOperation allows taking the average of a vector of tensors and enables gradient to flow through the average

The average operation will be useful in computing losses across a batch of data.

### CODE REVIEW

File: operations/averageOperation.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+#include "operations/operation.h"
+
+#ifndef __OP_AVG_INCLUDED__   
+#define __OP_AVG_INCLUDED__   
+
+template <typename T>
+class AverageOperation : public Operation<T> {
+    public:
+
+    vector<Tensor<T*>> tensors;
```
Comment (karanchahal) [inline operations/averageOperation.h:15]: Thanks for the pull request!! This is a prelim review.

The representation of the Average Tensor will need to be changed I think. Generally how the network takes in a tensor is a single tensor with size [batch_size, input_dim]. SO the average operation should take as input t1 which would be a tensor and another int variable called axis specifying what dimension to average along.  

Hence the API would look like

```c++
loss = losses::mse(y,gt)
averageLoss = tensorOps::average(loss,axis=0)
// and then continue from here
```

This would be inline how pytorch, tf and numpy handle some operation along a list of objects. Hence, I think it would be better to adhere with their API.

**Review by karanchahal — CHANGES_REQUESTED** [review]
Prelim Review.

File: operations/subtractOperation_impl.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    This file contains the implementation of the forward and backward pass of
+    the subtract operation.
+*/
+
+#include "operations/subtractOperation.h"
+
+#ifndef __OP_SUBTRACT_IMPL_INCLUDED__   
+#define __OP_SUBTRACT_IMPL_INCLUDED__  
+
+template <typename T>
```
Comment (karanchahal) [inline operations/subtractOperation_impl.h:11]: Comments on top of the function. Can remove comments inside function

File: overloads/tensor.h [hunk 3]
```diff
@@ -76,6 +119,21 @@ namespace tensorOps {
         return one->frontOp->forward();
     }
 
+    // Average
+    template<typename T>
+    Tensor<T>* average(vector<Tensor<T>*>& tensors) {
+        if (tensors.size() == 0) {  // To avoid division by zero. Should we do this?
```
Comment (karanchahal) [inline overloads/tensor.h:125]: Don't think this check will be needed if we adhere to the requested API

**Review by karanchahal — CHANGES_REQUESTED** [review]

File: operations/averageOperation.h (hunk 1 above)
Comment (uditarora) [reply operations/averageOperation.h:15]: Right, I missed the big picture here. Thanks for pointing this out.

Comment (karanchahal) [reply operations/averageOperation.h:15]: No problem, looking forward to your updates :)
